relay-server/src/metrics_extraction/metrics_summary.rs

```diff
@@ -0,0 +1,511 @@
+            BucketValue::Distribution(distribution) => Self::from_distribution(distribution),
+            BucketValue::Set(set) => Self::from_set(set),
+            BucketValue::Gauge(gauge) => Self::from_gauge(gauge),
+        }
+    }
+
+    /// Builds a [`MetricsSummaryBucketValue`] from a [`CounterType`].
+    fn from_counter(counter: &CounterType) -> MetricsSummaryBucketValue {
+        MetricsSummaryBucketValue {
+            min: Some(*counter),
+            max: Some(*counter),
+            sum: Some(*counter),
+            count: 1,
+        }
+    }
+
+    /// Builds a [`MetricsSummaryBucketValue`] from a [`DistributionValue`].
+    fn from_distribution(distribution: &DistributionValue) -> MetricsSummaryBucketValue {
+        let mut min = FiniteF64::MAX;
+        let mut max = FiniteF64::MIN;
+        let mut sum = FiniteF64::new(0.0).unwrap();
+
+        for value in distribution {
+            min = std::cmp::min(min, *value);
+            max = std::cmp::max(max, *value);
+            sum = sum.saturating_add(*value);
+        }
+
+        MetricsSummaryBucketValue {
+            min: Some(min),
+            max: Some(max),
+            sum: Some(sum),
+            count: distribution.len() as u64,
+        }
+    }
+
+    /// Builds a [`MetricsSummaryBucketValue`] from a [`SetValue`].
+    fn from_set(set: &SetValue) -> MetricsSummaryBucketValue {
+        // For sets, we limit to counting the number of occurrences.
+        MetricsSummaryBucketValue {
+            min: None,
+            max: None,
+            sum: None,
+            count: set.len() as u64,
+        }
+    }
+
+    /// Builds a [`MetricsSummaryBucketValue`] from a [`GaugeValue`].
+    fn from_gauge(gauge: &GaugeValue) -> MetricsSummaryBucketValue {
+        MetricsSummaryBucketValue {
+            min: Some(gauge.min),
+            max: Some(gauge.max),
+            sum: Some(gauge.sum),
+            count: gauge.count,
+        }
+    }
+
+    /// Merges two [`MetricsSummaryBucketValue`]s together by mutating `self` and consuming
+    /// `other`.
+    fn merge(&mut self, other: MetricsSummaryBucketValue) {
+        self.min = std::cmp::min(self.min, other.min);
+        self.max = std::cmp::max(self.max, other.max);
+        self.sum = match (self.sum, other.sum) {
+            (Some(sum), Some(other_sum)) => Some(sum.saturating_add(other_sum)),
+            (None, Some(other_sum)) => Some(other_sum),
+            (Some(sum), None) => Some(sum),
+            _ => None,
+        };
+        self.count += other.count;
+    }
+}
+
+/// metrics_summary_spec that tracks all the buckets containing the summaries for each
+/// [`MetricsSummaryBucketKey`].
+///
+/// The need for an metrics_summary_spec arises from the fact that we want to compute metrics summaries
+/// generically on any slice of [`Bucket`]s meaning that we need to handle cases in which
+/// the same metrics as identified by the [`MetricsSummaryBucketKey`] have to be merged.
```

Comment:
This doesn't explain what this is, requirements will change or the struct will be used elsewhere -> that comment does not apply anymore.

The comment should describe what this struct does and represents, not why it was created.